libgadget/blackhole.c

```diff
@@ -185,8 +205,61 @@ static double blackhole_soundspeed(double entropy, double rho) {
     return cs;
 }
 
+static void
+collect_BH_info(int * ActiveParticle,int NumActiveParticle, struct BHPriv *priv, FILE * FdBlackholeDetails)
+{
+    int i;
+    int c=0;
+    
+    for(i = 0; i < NumActiveParticle; i++)
+    {
+        int p_i = ActiveParticle ? ActiveParticle[i] : i;
+        
+        if(P[p_i].Type != 5 || P[p_i].IsGarbage || P[p_i].Mass <= 0)
+          continue;
+        
+        int PI = P[p_i].PI;
+        
+        struct BHinfo info;
+        info.ID = P[p_i].ID;
+        info.Mass = BHP(p_i).Mass;
+        info.Mdot = BHP(p_i).Mdot;
+        info.Density = BHP(p_i).Density;
+        info.minTimeBin = BHP(p_i).minTimeBin;
+        info.MinPotPos[0] = BHP(p_i).MinPotPos[0];
+        info.MinPotPos[1] = BHP(p_i).MinPotPos[1];
+        info.MinPotPos[2] = BHP(p_i).MinPotPos[2];
+        
+        info.MinPot = priv->MinPot[PI];
+        info.BH_Entropy = priv->BH_Entropy[PI];
+        info.BH_SurroundingGasVel[0] = priv->BH_SurroundingGasVel[PI][0];
+        info.BH_SurroundingGasVel[1] = priv->BH_SurroundingGasVel[PI][1];
+        info.BH_SurroundingGasVel[2] = priv->BH_SurroundingGasVel[PI][2];
+        
+        info.BH_accreted_BHMass = priv->BH_accreted_BHMass[PI];
+        info.BH_accreted_Mass = priv->BH_accreted_Mass[PI];
+        info.SPH_SwallowID = priv->SPH_SwallowID[PI];
+        
+        info.a = All.Time;
+        
+        int size = sizeof(info);
+        
+        fwrite(&size, sizeof(size), 1, FdBlackholeDetails);
+        fwrite(&info,sizeof(info),1,FdBlackholeDetails);
+        fwrite(&size, sizeof(size), 1, FdBlackholeDetails);
```

Comment:
As I said to Yu last time this came up, I'm not a fan of the extra binary file format. But we can merge and port to the BigFileRecords later.

Reply:
Good old fortran record format